packages/keystone/adapterCache.js

```diff
@@ -0,0 +1,514 @@
+            // Useful for hotfixes or disabling cache for business critical lists
+            this.excludedLists = excludedLists || []
+
+            // This mechanism allows to control garbage collection.
+            this.maxCacheKeys = maxCacheKeys || 1000
+
+            // Cache: { listName -> queryKey -> { response, lastUpdate, score } }
+            this.cache = new LRUCache({ max: this.maxCacheKeys })
+
+            // Logging allows to get the percentage of cache hits
+            this.logging = !!logging
+
+            this.totalRequests = 0
+            this.cacheHits = 0
+
+        } catch (e) {
+            this.enabled = false
+            logger.warn(`ADAPTER_CACHE: Bad config! reason ${e}`)
+        }
+    }
+
+    /**
+     * Sometimes we might get a situation when two instances try to update state with different timestamp values.
+     * So, we might have two commands ordered to execute by internal redis queue:
+     * 1. SET some_key 1670000000010
+     * 2. SET some_key 1670000000009
+     * In result (GET some_key) will equal 1670000000009, but the correct value is 1670000000010!
+     * So to counter this we write custom redis function that will update value only if it is bigger!
+     */
+    getStateRedisClient () {
+        const updateTimeStampFunction = {
+            numberOfKeys: 1,
+            lua: `
+            local time = tonumber(ARGV[1])
+            local old_time = tonumber(redis.call('GET', KEYS[1]))
+            if (old_time == nil) then
+                return redis.call('SET', KEYS[1], ARGV[1])
+            end
+            if (time > old_time) then
+                return redis.call('SET', KEYS[1], ARGV[1])
+            end
+        ` }
+
+        const redis = getRedisClient('cache')
+        redis.defineCommand('cacheUpdateStateTimestamp', updateTimeStampFunction)
+        return redis
+    }
+
+    /**
+     * Sets last updated list time to Redis storage
+     * @param {string} listName -- List name
+     * @param {Date} time -- Last updated time
+     * @returns {Promise<void>}
+     */
+    async setState (listName, time) {
+        const serializedTime = time.valueOf()
+        const prefixedKey = `${STATE_REDIS_KEY_PREFIX}:${listName}`
+        await this.redisClient.cacheUpdateStateTimestamp(prefixedKey, serializedTime)
+    }
+
+    /**
+     * Returns last updated time by list from Redis
+     * @param {string} listName -- List name
+     * @returns {Promise<Date>}
+     */
+    async getState (listName) {
+        const serializedTime = await this.redisClient.get(`${STATE_REDIS_KEY_PREFIX}:${listName}`)
+        if (serializedTime) {
+            const parsedTime = parseInt(serializedTime)
+            if (!isNaN(parsedTime)) { return new Date(parsedTime) }
+        }
+        return null
+    }
+
+    /**
+     * Set value to cache
+     * @param {string} listName
+     * @param {string} key
+     * @param {Object} value
+     */
+    setCache (listName, key, value) {
```

Comment:
It look too comples: setState and setCache. Two similar function with a little difference